lib/spack/docs/build_systems/gopackage.rst

```diff
@@ -0,0 +1,164 @@
+        version('0.11.3', sha256='b3c503de6b34fd4284fd70655e7f42feafc07f090e7f7cc00db261f56c583c46')
+
+        depends_on('go@1.13:', type='build')  # go.mod value overrides default
+
+        executables = ['lazygit']
+
+While it generally follows the common patterns of a Spack package,
+it's worth calling attention to its:
+
+* adding a *build* dependency on the version of the Go package that it
+  requires (from the existing `go.mod`);
+* using the default ``build`` and ``install`` steps that ``GoPackage``
+  provides; and
+* providing the name of the executable to be installed by the default
+  ``install`` step.
+
+The ``GoPackage`` provides a `setup_build_environment` that sets
+`GO111MODULES=on` and `GOFLAGS=-mod=vendor` and its default build step
+invokes ``go build``.  Additional arguments to ``go build`` can be
+provided by setting ``build_args``, e.g. one could
+
+.. code-block:: python
+
+        build_args = ['-p', '12']  # adjust parallelism of build cmd
+
+to invoke ``go build -p 12``.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Slightly more complicated packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Packages that can be built with a simple invocation of ``go build``,
+that use modules, *but have not vendored their dependencies* require a
+bit of extra work.  In particular, in order to satisfy Spack's
+requirement that nothing is downloaded during the build step, the
+package needs to define "resources" that describe from whence to fetch
+each dependency, where to emplace it, and the version of the package
+to which it applies.  It is possible (likely) that each release of the
+project will new set of resources definitions covering the correct
+versions.
+
+In the general case, compiling these resource is complicated.
+
+Fortunately, packagers of projects that use modules can use an
+automated tool (``modules2tuple``) to provide the resource definitions
+and ``import_resources`` to include them without overly-complicating
+the package definition.  ``modules2tuple``
+(https://github.com/dmgk/modules2tuple) was originally developed by
+the FreeBSD Ports team to generate the Makefile fragment their build
+system uses to represent a Go application's dependencies.  For the
+curious, their Porters Handbook includes more details about `packaging
+Go Applications
+<https://www.freebsd.org/doc/en_US.ISO8859-1/books/porters-handbook/building.html#using-go>`_
+for FreeBSD.
+
+There are three steps to describing the dependencies in the package
+definition:
+
+* First, stage the application and use the Go tool chain to
+  solve the set of required dependencies and build a vendor directory.
+
+* Second, run the newly created ``vendor/modules.txt`` file through
+  ``module2tuples`` to generate a JSON file containing resource
+  definitions.
+
+.. code-block:: console
+
+    # in the root directory of the project (where the go.mod file lives)
+    $ go mod vendor
+    $ modules2tuple -spack -app_version=1.2.3 vendor/modules.txt > resources-1.2.3.json
```

Comment:
Did you add the `-spack` flag to this tool yourself?

Reply:
Yes, see [the PR](https://github.com/dmgk/modules2tuple/pull/8).  I spent a bunch of time looking at how other projects handle Go applications and FreeBSD's feels cleanest.   Go provides some neat indirection between the name of imports in the code and where it actually needs to be fetched from, and then there's its [revolutionary approach to version selection](https://research.swtch.com/vgo-mvs) that is intended to/seems to/should make long term support of code less painful.  They provide really nice tooling for checksumming and mirroring/proxying and etc..., but it's all Go specific and there aren't any really nice places to hook in other systems that want to handle those bits (aka, us).

There's some awkward stuff (e.g. vanity imports and mirrors) that needs to happen and I didn't want to leave the Spack project having to support a transliteration tool on its own.  Drafting (in the bicycle sense) the FreeBSD team seems easiest.